thunder/distributed/__init__.py

```diff
@@ -177,6 +177,102 @@ def prep_shard(
+    broadcast_from=None,
+    bucket_size_in_mb: float = 25.0,
+) -> ThunderModule:
+    from thunder import compile_data as get_compile_data
+    from thunder.core.transforms import add_transform
+    from thunder.core.module import ThunderModule
+    from thunder.distributed.transforms.ddp_v2 import DDPTraceTransform
+
+    process_group = copy_default_process_group()
+    utils.check(process_group is not None, lambda: "The default process group is None")
+    cd = get_compile_data(thunder_model)
+    cd.use_ddp = True
+    orig_module: torch.nn.Module = cd.fn
+    utils.check(
+        isinstance(orig_module, torch.nn.Module) and not isinstance(orig_module, ThunderModule),
+        lambda: f"CompileData.fn expected to be `nn.Module` but {type(orig_module)}",
+    )
+    orig_module.use_ddp = True
+    orig_module.process_group_for_ddp = process_group
+    orig_module.bucket_size_in_mb = bucket_size_in_mb
+
+    replicated_params = {}
+    # We use `shared_params` dictionary to track the shared parameters.
+    # Key to this dictionary is the original parameter from the user's Module.
+    # Values are the copied and sharded parameter for the thunder module and meta-data related to sharding.
+    shared_params = WeakTensorKeyDictionary()
+
+    # NOTE: Shared Parameters in Trace
+    # Shared parameters in PyTorch eager are parameters of module which have different name but share the underlying tensor.
+    # For shared parameter, we replace all occurence shared parameter with it's corresponding `base` parameter.
+    # In our implementation `base` parameter is the parameter and corresponding name which we see the first time while
+    # iterating our parameters (see below). We track subsequent parameter which share the underlying Tensor with this `base` parameter
+    # in `shared_params_name` dictionary.
+    # Then while, transforming the trace - `see DDPTraceTransform.transform_traces` - we replace all the proxy of shared parameter
+    # with the corresponding proxy of base parameter in the computation trace.
+
+    # This is used to track the shared parameters when the transform is applied.
+    # key - parameter name, value - `base` parameter name.
+    shared_params_name: dict[str, str] = {}
+    for module_name, _ in thunder_model._model.named_modules():
+        submodule = thunder_model.get_submodule(module_name)
+        # Since we are doing no sharding, we do not need to materialize the params
+
+        # Broadcast parameters if requested
+        if broadcast_from is not None:
+            for pn, _ in submodule.named_parameters(recurse=False, prefix=module_name):
+                tdist.broadcast(
+                    thunder_model.get_parameter(pn), src=broadcast_from, group=process_group, async_op=False
+                )
+            for pn, _ in submodule.named_buffers(recurse=False, prefix=module_name):
+                tdist.broadcast(thunder_model.get_buffer(pn), src=broadcast_from, group=process_group, async_op=False)
```

Comment:
PyTorch's `DistributedDataParallel` seems to broadcast params always as per https://github.com/pytorch/pytorch/blob/784a6ec5a30bd2d1831cb4f78183ad51696794e5/torch/nn/parallel/distributed.py#L823-L831

Reply:
> PyTorch's `DistributedDataParallel` seems to broadcast params always as per https://github.com/pytorch/pytorch/blob/784a6ec5a30bd2d1831cb4f78183ad51696794e5/torch/nn/parallel/distributed.py#L823-L831

The thunder.distributed.ddp(...) function (the user API) takes a broadcast_from that defaults to 0, but can be set to None to avoid broadcasting. In the PyTorch API, it always seems to broadcast parameters to 0. Isn't it better to let the user specify if they don't want to broadcast (for any weird reason)?.

Reply:
uh... I seem to have been confused.
We allow `broadcast_from=None` to skip broadcast of ddp, but the default value is `0` as per https://github.com/Lightning-AI/lightning-thunder/blob/67ec2c6f4e72f6215f12941bf161915f8f458181/thunder/distributed/__init__.py#L188.
okay, no changes needed here :)